### PR TITLE
Add a table for compliance results

### PR DESCRIPTION
**Pull request by rhmdnd**
This commit adds the basic schema we've outlined in
https://github.com/rhmdnd/compserv/issues/84#issuecomment-1182304033
Future patches might include additional columns we need, depending on
how this is used.

Fixes #31

### CODE REVIEW

File: migrations/000002_create_results_table.up.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE IF NOT EXISTS results (
+        id uuid PRIMARY KEY,
+        name VARCHAR(255),
+        control VARCHAR(255),
+        outcome VARCHAR(255),
+        instruction TEXT,
+        metadata_id UUID,
+        subject_id UUID,
+        CONSTRAINT fk_results_metadata_id FOREIGN KEY (metadata_id) REFERENCES metadata (id),
```
Comment (rhmdnd) [inline migrations/000002_create_results_table.up.sql:9]: The migration CI will fail here since we haven't merged the metadata table, yet.

File: migrations/000002_create_results_table.up.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE IF NOT EXISTS results (
+        id uuid PRIMARY KEY,
+        name VARCHAR(255),
+        control VARCHAR(255),
+        outcome VARCHAR(255),
+        instruction TEXT,
```
Comment (rhmdnd) [inline migrations/000002_create_results_table.up.sql:6]: The `TEXT` type is a pretty big hammer. This could be abused to bloat the database.

File: tests/utils.go [hunk 3]
```diff
@@ -96,3 +106,35 @@ func getUUIDString() string {
 	value, _ := uuid.NewRandom()
 	return value.String()
 }
+
+func insertMetadata() (string, error) {
```
Comment (rhmdnd) [inline tests/utils.go:110]: While this is useful to have in this patch, I think it could be introduced in a separate patch.

File: tests/utils.go [hunk 4]
```diff
@@ -96,3 +106,35 @@ func getUUIDString() string {
...
+
+	md := Metadata{ID: id, CreatedAt: createdAt, UpdatedAt: updatedAt, Version: version, Description: description}
+	if err := gormDB.Create(&md).Error; err != nil {
+		return "", err
+	}
+
+	return id, nil
+}
+
+func insertSubject() (string, error) {
```
Comment (rhmdnd) [inline tests/utils.go:127]: Ditto

**Comment by rhmdnd** [comment]
I think we should hold off on merging this until we add a schema for assessments, which is being discussed https://github.com/rhmdnd/compserv/issues/118

Primary reasoning is that results will need to link to an assessment, so we'll need to update this schema to include another foreign key to the assessment table.

File: migrations/000005_create_results_table.up.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE IF NOT EXISTS results (
```
Comment (Vincent056) [inline migrations/000005_create_results_table.up.sql:1]: Sorry for the late review, I am wondering if we should match the compliance check result CRD,
the schema we have right now is missing `Warnings`, `Description`, `Severity`, and `ValuesUsed`.
Are we planning to store those and keep them somewhere, or do we not care about these data?

File: migrations/000002_create_results_table.up.sql (hunk 2 above)
Comment (jhrozek) [reply migrations/000002_create_results_table.up.sql]: Could we limit the size to something sane as part of the API or is that considered bad practice?

File: migrations/000005_create_results_table.up.sql (hunk 5 above)
Comment (rhmdnd) [reply migrations/000005_create_results_table.up.sql:1]: That's a good question. `Description` can be used with the `metadata_id` column and point to an entry there.

https://github.com/rhmdnd/compserv/blob/main/migrations/000002_create_metadata_table.up.sql#L6

Is `Severity` something that is specific to results, or is also applicable to controls? Should it be in `metadata`? I'm unsure what to do with `Warnings` and `ValuesUsed`.

File: migrations/000002_create_results_table.up.sql (hunk 2 above)
Comment (rhmdnd) [reply migrations/000002_create_results_table.up.sql]: I'd argue having consistent API validation is good practice, because it reduces the chance that a change in backend will have a impact on the API.

**Comment by jhrozek** [comment]
On Wed, Aug 03, 2022 at 06:29:09AM -0700, Lance Bragstad wrote:
> @rhmdnd commented on this pull request.
> 
> 
> 
> > @@ -0,0 +1,13 @@
> +CREATE TABLE IF NOT EXISTS results (
> 
> That's a good question. `Description` can be used with the `metadata_id` column and point to an entry there.
> 
> https://github.com/rhmdnd/compserv/blob/main/migrations/000002_create_metadata_table.up.sql#L6
> 
> Is `Severity` something that is specific to results, or is also applicable to controls?

IIRC result's severity is inherited from control's severity.

> Should it be in `metadata`?

Isn't metatada used by both subjects and assessments as well? Those
don't have a severity, do they?

> I'm unsure what to do with `Warnings` and `ValuesUsed`.

I see some value in warnings, but at the same time I'm unsure how (if) the
consumer would consume them.
Same with ValuesUsed, sounds useful.

All in all, since these would be just about adding additional data to
the result, I guess we could add them later. What seems the be important
is to design the tables in a way that we could support at least the queries we
suspect that the consumer of the service would be doing.

**Comment by rhmdnd** [comment]
> Isn't metatada used by both subjects and assessments as well? Those don't have a severity, do they?

I feel like `severity` makes sense for the control. If we associate `severity` to the result, and multiple rules or results are used to check a control, which severity takes precedence? Probably the one with the higher value (e.g., `high` over `medium`), but we might be introducing a precedence problem if we associate `severity` to the result and if we can have multiple results associated to a single control.

**Comment by rhmdnd** [comment]
@jhrozek @Vincent056 I added two additional schemas. One for profiles and another for controls to https://github.com/rhmdnd/compserv/issues/84

The general idea @Vincent056 and I discussed is that the result would only store a reference/ID to the control. Then, the compliance service would derive all control information from the source (OSCAL definitions).

We ended up here because it wouldn't duplicate control information in the result table. Additionally, we tried to be as DRY as possible when storing control and profile information.

File: migrations/000006_create_results_table.up.sql [hunk 6]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE IF NOT EXISTS results (
+        id uuid PRIMARY KEY,
+        name VARCHAR(255),
+        control VARCHAR(255),
+        outcome VARCHAR(255),
+        instruction TEXT,
```
Comment (jhrozek) [inline migrations/000006_create_results_table.up.sql:6]: based on what we talked about yesterday, we're going to use all the CO's attributes and store them as a fallback right?

**Review by jhrozek — APPROVED** [review]
/lgtm
